presto-spi/src/main/java/com/facebook/presto/spi/block/Block.java

```diff
@@ -193,12 +196,15 @@ default int compareTo(int leftPosition, int leftOffset, int leftLength, Block ri
 
     /**
      * Create a new block from the current block by keeping the same elements
-     * only with respect to {@code visiblePositions}.
+     * only with respect to {@code positions} that starts at {@code offset} and has length of {@code length}.
      * May return a view over the data in this block or may return a copy
      */
-    default Block getPositions(int[] positions)
+    default Block getPositions(int[] positions, int offset, int length)
     {
-        return new DictionaryBlock(this, positions);
+        checkValidPositionsArray(positions, offset, length);
+
+        int[] ids = length == positions.length ? positions : Arrays.copyOfRange(positions, offset, offset + length);
```

Comment:
Not very readable.

Also, did you consider `DictionaryBlock` constructor that takes `idsOffset` instead of copying the array?

Reply:
+1, please use: `DictionaryBlock#DictionaryBlock(int, int, com.facebook.presto.spi.block.Block, int[], boolean, com.facebook.presto.spi.block.DictionaryId)`

Reply:
That's a good point. Two questions:
- That requires the interface to be public. Hope that is minor.
- Do we wanna check `compact` in this case? Checking can be quite expensive. Or by default we just make it `false`?

Reply:
Good questions. Not my call though.